www/components/MyST.tsx

```diff
@@ -0,0 +1,270 @@
+import * as md from "mdast";
+import { MyST } from "mystjs";
+import React, { ReactElement, useContext } from "react";
+import { Literal, Node } from "unist";
+
+export interface MySTMarkupProps extends MySTRoles {
+  source: string;
+}
+
+export interface MySTRoles {
+  roles?: (role: MySTRole) => ReactElement;
+}
+
+const MySTContext = React.createContext<MySTRoles>({ roles: (_) => <></> });
+
+export default function MySTMarkup({ source, roles }: MySTMarkupProps) {
+  const parser = new MyST();
+  const tree = parser.parse(source);
+  return (
+    <MySTContext.Provider value={{ roles }}>
+      <MySTParent node={tree} />
+    </MySTContext.Provider>
+  );
+}
+
+interface DispatchProps<T extends Node> {
+  key: number;
+  node: T;
+}
+
+export interface MySTRole extends Literal {
+  type: "mystRole";
+  name: string;
+}
+
+const dispatch = {
```

Comment:
Nnngh :D I've spent ~20 minutes trying to come up with something better, but failed. It feels like there should be a way to use the `mdast` field `type` as a discriminator between the various `mdast` types *in the type system* instead of a literal runtime lookup in a map, but I'm not sure how to do that without it being like 4x as verbose as this. So... /shrug

(Really wish TS had proper function overloading / dynamic dispatch over types)

Reply:
Ooh, ooh, https://github.com/syntax-tree/unist#list-of-utilities?

Reply:
Alright, this is my best idea at making TS here happy and also maximizing the safety we get from the type system: https://gist.github.com/abesto/6ef437da8ab8568e7b757c22455d58f6